test/performance/pacers.go

```diff
@@ -24,38 +24,55 @@ import (
 	vegeta "github.com/tsenart/vegeta/lib"
 )
 
-// SteadyUpPacer is a Pacer that describes attack request rates that increases in the beginning then becomes steady.
+// steadyUpPacer is a Pacer that describes attack request rates that increases in the beginning then becomes steady.
 //  Max  |     ,----------------
 //       |    /
 //       |   /
 //       |  /
 //       | /
 //  Min -+------------------------------> t
 //       |<-Up->|
-type SteadyUpPacer struct {
+type steadyUpPacer struct {
 	// UpDuration is the duration that attack request rates increase from Min to Max.
 	UpDuration time.Duration
-	// Min is the attack request rates from the beginning.
+	// Min is the attack request rates from the beginning, must be larger than 0.
 	Min vegeta.Rate
 	// Max is the maximum and final steady attack request rates.
 	Max vegeta.Rate
+
+	slope        float64
+	minHitsPerNs float64
+	maxHitsPerNs float64
+}
+
+// NewSteadyUpPacer returns a new SteadyUpPacer with the given config.
+func NewSteadyUpPacer(min vegeta.Rate, max vegeta.Rate, upDuration time.Duration) vegeta.Pacer {
+	pacer := &steadyUpPacer{
+		Min:        min,
+		Max:        max,
+		UpDuration: upDuration,
+	}
+	pacer.slope = (hitsPerNs(max) - hitsPerNs(min)) / float64(upDuration)
+	pacer.minHitsPerNs = hitsPerNs(pacer.Min)
+	pacer.maxHitsPerNs = hitsPerNs(pacer.Max)
+	return pacer
 }
 
-// SteadyUpPacer satisfies the Pacer interface.
-var _ vegeta.Pacer = SteadyUpPacer{}
+// steadyUpPacer satisfies the Pacer interface.
+var _ vegeta.Pacer = steadyUpPacer{}
 
-// String returns a pretty-printed description of the SteadyUpPacer's behaviour.
-func (sup SteadyUpPacer) String() string {
+// String returns a pretty-printed description of the steadyUpPacer's behaviour.
+func (sup steadyUpPacer) String() string {
 	return fmt.Sprintf("Up{%s + %s / %s}, then Steady{%s}", sup.Min, sup.Max, sup.UpDuration, sup.Max)
 }
 
-// invalid tests the constraints documented in the SteadyUpPacer struct definition.
-func (sup SteadyUpPacer) invalid() bool {
-	return sup.UpDuration <= 0 || hitsPerNs(sup.Min) <= 0 || hitsPerNs(sup.Max) <= hitsPerNs(sup.Min)
+// invalid tests the constraints documented in the steadyUpPacer struct definition.
+func (sup steadyUpPacer) invalid() bool {
+	return sup.UpDuration <= 0 || sup.minHitsPerNs <= 0 || sup.maxHitsPerNs <= sup.minHitsPerNs
```

Comment:
this feels like it belongs in `New`

Reply:
Same as above, this function is defined to follow the same patterns as the existed pacers:
Stop the attack as soon as it finds the pacer to be `invalid` in the `Pace` function.

Reply:
Yeah, but you shouldn't even be able to create a pacer that's `invalid`, since none of the parameters are actually set after creation (which I think is different from vegeta pacers).